### PR TITLE
Fix unresolved type span inside async object

### LINKED ISSUE #65180: Async-related type error messages defy expectations (in span location)
Consider for example this code (using `hyper-0.13.0`):

```rust
use std::net::{Ipv4Addr, Ipv6Addr, SocketAddr, SocketAddrV4, SocketAddrV6, TcpListener};

#[derive(failure::Fail, Debug)]
enum Error {
    #[fail(display="Cannot bind to port {}", _1)]
    CannotBind(#[cause] std::io::Error, u16),
    #[fail(display="Cannot create a server")]
    UncreatableServer(#[cause] hyper::Error),
    #[fail(display="Could not start serving metrics")]
    Unservable(#[cause] hyper::Error),
}

async fn serve_metrics(
    port: u16,
) -> Result<(), Error> {
    let bind_addrs = [
        SocketAddr::V6(SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, port, 0, 0)),
        SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, port)),
    ];
    let listener = TcpListener::bind(&bind_addrs[..]).map_err(|e| Error::CannotBind(e, port))?;
    let server = hyper::server::Server::from_tcp(listener).map_err(Error::UncreatableServer)?;

    let server = server.serve(hyper::service::make_service_fn(move |_| {
        async move {
            Ok(hyper::service::service_fn(move |req| {
                async move {
                    hyper::Response::builder()
                        .status(hyper::StatusCode::NOT_FOUND)
                        .body(hyper::Body::empty())
                }
            }))
        }
    }));
    server.await.map_err(Error::Unservable)
}
```

This code will fail to compile due to types which cannot be fully inferred, with an error like this:

```
error[E0698]: type inside `async` object must be known in this context
  --> src/lib.rs:23:25
   |
23 |     let server = server.serve(hyper::service::make_service_fn(move |_| {
   |                         ^^^^^ cannot infer type for `ME`
   |
note: the type is part of the `async` object because of this `await`
  --> src/lib.rs:34:5
   |
34 |     server.await.map_err(Error::Unservable)
   |     ^^^^^^^^^^^^
```

However neither of these spans are anywhere near to the real problem which is the fact that `Err` variant is not inferred for `Ok(hyper::service::service_fn(...))`.

### PR DESCRIPTION
**Pull request by csmoe**
Closes https://github.com/rust-lang/rust/issues/65180
r? @estebank 
cc @nikomatsakis @cramertj

### CODE REVIEW

**Comment by estebank** [comment]
Does this change have a user visible change?

**Comment by csmoe** [comment]
@estebank 
From
```txt


error[E0698]: type inside `async` object must be known in this context
  --> src/lib.rs:23:25
   |
23 |     let server = server.serve(hyper::service::make_service_fn(move |_| {
   |                         ^^^^^ cannot infer type for `ME`
   |
note: the type is part of the `async` object because of this `await`
  --> src/lib.rs:34:5
   |
34 |     server.await.map_err(Error::Unservable)
   |     ^^^^^^^^^^^^
```
To:
```txt
error[E0698]: type inside `async fn` body must be known in this context
  --> src/lib.rs:15:13
   |
15 |             Ok(hyper::service::service_fn(move |req| {
   |             ^^ cannot infer type for `ME`
   |
note: the type is part of the `async fn` body because of this `await`
  --> src/lib.rs:24:5
   |
24 |     server.await;
   |     ^^^^^^^^^^^^

error: aborting due to previous error
```
but I should figure out a test/ui.

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -1833,6 +1833,16 @@ impl fmt::Display for YieldSource {
     }
 }
 
+impl core::convert::From<GeneratorKind> for YieldSource {
```
[the remark this answers is not in the document]
Comment (sinkuu) [reply src/librustc/hir/mod.rs]: Also, `From` is included in [prelude](https://doc.rust-lang.org/std/prelude/v1/index.html), right?

```suggestion
impl From<GeneratorKind> for YieldSource {
```

**Comment by estebank** [comment]
r=me with a test

**Comment by nikomatsakis** [comment]
Looks good to me, too

**Comment by JohnCSimon** [comment]
Ping from triage.
@estebank can you please review this PR?
cc: @nikomatsakis @csmoe 
Thank you!

**Comment by estebank** [comment]
@csmoe can you add a ui test exercising this new code?

**Comment by csmoe** [comment]
@estebank it's a bit complicated, still trying.

**Comment by JohnCSimon** [comment]
Ping from triage:
@csmoe How's it going with the UI test? 
cc: @estebank 
Thank you!

**Comment by JohnCSimon** [comment]
Ping from triage:
@csmoe Can you please post your status regarding this PR, and address the merge conflicts?
cc: @estebank
Thank you!

**Comment by JohnCSimon** [comment]
Pinging once again from triage:
@csmoe Can you please post your status regarding this PR, and address the merge conflicts?

Thank you.

**Comment by JohnCSimon** [comment]
Triage: 

@csmoe - unfortunately this PR has sat idle for several weeks, I'm marking this as inactive. Please reopen this PR when you're ready to move forward. 
Thank you

**Pull request closed without merging**